src/lib.rs

```diff
@@ -1,4 +1,99 @@
+//!     hstart: *const usize,
+//!     hptr: *mut usize,
+//!
+//!     // Flag to determine whether to collect at the next safepoint
+//!     collect_next: bool,
+//!
+//!     // The in-memory safepoint table used to identify roots
+//!     roots: HashMap<ReturnAddress, SafepointRoots>
+//! }
+
+#[cfg(not(all(target_pointer_width = "64", target_arch = "x86_64")))]
 compile_error!("Requires x86_64 with 64 bit pointer width.");
 
-pub mod safepoints;
+mod collector;
+mod safepoints;
+use collector::Collector;
+
+// FIXME: This will be replaced with the `Scan` trait lang item in our forked
+// rustc's libcore. For now, we define `Scan` at the top level in this library.
+pub trait Scan {
+    fn scan(&self);
+}
+
+thread_local!(static COLLECTOR: Collector =  Collector::new());
+
+/// This must be called before the GC can be used (usually in the setup code
+/// before `main()`). Initialisation consists of two stages:
+///     1. Read the stackmap section in the ELF file into an in-memory table for
+///        fast lookup.
+///     2. Allocate a chunk of heap memory to be used to store objects managed
+///        by the GC.
+pub fn init() {
+    unimplemented!();
+}
+
+/// This function is the *only* way that a collection can be triggered. Calls to
+/// `safepoint_poll` are generated by LLVM's InsertSafepoints opt pass. They are
+/// inserted liberally into the mutator's code at all function calls and
+/// loop-backedges.
+///
+/// In almost all cases, a safepoint poll will *never* trigger a collection.
+/// It's therefore really important that this function returns fast on a
+/// wont-collect poll.
+///
+/// No information about whether a poll resulted in a collection is returned to
+/// the mutator. The only thing that can be guaranteed is that a collection
+/// *might* have happened after returning from this call.
+///
+/// ------------------------------ WARNING -------------------------------------
+/// | We need to be really careful about code which panics inside the collector|
+/// | here. The safepoint poll will *not* be called by native Rust code which  |
+/// | means that panic handling is UB. We should probably abort on panic here. |
+/// ----------------------------------------------------------------------------
+#[no_mangle]
+pub extern "C" fn safepoint_poll() {
+    if COLLECTOR.with(|c| c.should_collect()) {
+        COLLECTOR.with(|c| c.reclaim())
+    }
+}
+
+/// Perform a collection at the next safepoint. Note that this *does not* stop
+/// the world immediately to perform garbage collection. Instead, it guarantees
+/// that once the next safepoint boundary has been reached, the collector will
+/// attempt a collection.
+///
+/// Calling `force_collect` multiple times before a safepoint boundary is
+/// pointless - it won't make a collection happen any faster. It simply sets a
+/// flag in the collector which is checked in order to decide whether to make a
+/// collection.
+pub fn force_collect() {
```

Comment:
I am not sure about `force_collect` -- because it doesn't force a collection AFAICS? Or does it block the current thread until a collection has occurred?

Reply:
I'm not sure it's the best name either. You're right, that it _doesn't_ force a collection immediately, but it means that the next time the mutator hits a safepoint, the collector will pause to reclaim memory. It won't block the current thread.

Reply:
Right, so we need to change the name to something that doesn't suggest that it forces a GC. I'd suggest something like `request_collect_soon` which is a bit of a mouthful, but at least not easily misunderstood.

Reply:
I've changed this to `request_collect_soon`, but I don't really like it. I'm not sure what would be better though.

Reply:
Maybe `queue_collect` or `schedule_collect` or similar?